### PR TITLE
Better support: Bitbucket server + Xcode server

### PR DESCRIPTION
**Pull request by krin-san**
While using DangerCI in my project setup I noticed a bunch of issues fixed with this PR:
* BitBucket Server has issues with default emojis _(I got HTTP 500 when tried to pos a comment with default emojis)_
* BitBucket Server was not listed as a supported source for XcodeServer
* Different emojis was used without mapping with `character_from_emoji`
* BitBucket Server [disable inline HTML syntax support](https://jira.atlassian.com/browse/BSERV-7147) in comments. [Current syntax](https://confluence.atlassian.com/bitbucketserver/markdown-syntax-guide-776639995.html#Markdownsyntaxguide-Tables) is limited

_p.s.: Ruby is not a language I know well so any advice about "how you shouldn't write in Ruby" could be helpful._

### CODE REVIEW

File: lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb [hunk 1]
```diff
@@ -150,7 +150,9 @@ def head_commit
     # @return [String]
     #
     def html_link(paths, full_path: true)
-      create_link(paths, full_path) { |href, text| create_html_link(href, text) }
+      # TODO: Enable html formatting when BitBucketServer could handle it
+      #create_link(paths, full_path) { |href, text| create_html_link(href, text) }
+      markdown_link(paths, full_path: full_path)
```
Comment (krin-san) [inline lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb]: Tests are failing at this point. I have two options how to fix them:
* completely remove html formatting code with tests
* move this fix to `danger-xcode_summary` plugin repo and call a proper method based on chosen RequestSource

File: lib/danger/helpers/comments_helper.rb [hunk 2]
```diff
@@ -151,17 +152,25 @@ def random_compliment
...
+          "bitbucket_server" => {
+            "no_entry_sign"    => "\u274C", # ❌
+            "warning"          => "\u2757", # ❗
+            "book"             => "\u2728", # ✨
+            "white_check_mark" => "\u2705"  # ✅
+          }
+        }
+
+        return container[template][emoji]
       end
```
Comment (JuanitoFatas) [inline lib/danger/helpers/comments_helper.rb]: What do you think about extract the logic here to a class?

```ruby
def character_from_emoji(emoji, template: "github")
  WarningEmoji.new(emoji, template).call
end
```

```ruby
class WarningEmoji
  DATA = {
    "github" => {
      "no_entry_sign"    => "🚫",
      "warning"          => "⚠️",
      "book"             => "📖",
      "white_check_mark" => "✅"
    },
    "bitbucket_server" => {
      "no_entry_sign"    => "\u274C", # ❌
      "warning"          => "\u2757", # ❗
      "book"             => "\u2728", # ✨
      "white_check_mark" => "\u2705"  # ✅
    }
  }

  def initialize(emoji, template)
    @emoji = emoji.delete ":"
    @template = template
  end

  def call
    DATA[template][emoji]
  end

  private

    attr_reader :emoji, :template
end
```

I am bad at naming things, you can change to what suits the best :)

File: lib/danger/helpers/comments_helper.rb [hunk 3]
```diff
@@ -134,11 +135,11 @@ def generate_inline_markdown_body(markdown, danger_id: "danger", template: "gith
         )
       end
 
-      def generate_description(warnings: nil, errors: nil)
+      def generate_description(warnings: nil, errors: nil, template: "github")
         if errors.empty? && warnings.empty?
           return "All green. #{random_compliment}"
         else
-          message = "⚠ "
+          message = character_from_emoji(":warning", template: template) + " "
```
Comment (JuanitoFatas) [inline lib/danger/helpers/comments_helper.rb]: Can we pass in `"warning"` directly?

File: lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb (hunk 1 above)
Comment (JuanitoFatas) [inline lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb]: > - completely remove html formatting code with tests

How about drop a comment and skip these tests for now? 😊

**Comment by JuanitoFatas** [comment]
Thank you so much for this 🙏 . Looks good 😍  with few comments 👍

File: spec/lib/danger/helpers/comments_helper_spec.rb [hunk 4]
```diff
@@ -258,7 +258,7 @@ class Dummy
     it "handles a single error and a single warning" do
       message = dummy.generate_description(warnings: [1], errors: [1])
 
-      expect(message).to include("⚠ ")
+      expect(message).to include("⚠️ ")
```
Comment (orta) [inline spec/lib/danger/helpers/comments_helper_spec.rb]: heh - tofu

File: lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb (hunk 1 above)
Comment (krin-san) [inline lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb]: @JuanitoFatas, offer a best way to skip these tests:
* using metadata `, skip: true`
* prefixing with `x` like `it -> xit`
* removing tests completely. This way we also do't see any "No reason given" warnings

File: lib/danger/helpers/comments_helper.rb (hunk 2 above)
Comment (krin-san) [inline lib/danger/helpers/comments_helper.rb]: Good idea, but it needs some refactoring.
Would be nice to have a separate attribute like `@emoji_mapper` to map emoji character names into chars without re-defining the template. How about this?
```
class EmojiMapper
  DATA = {
    "github" => {
      "no_entry_sign"    => "🚫",
      "warning"          => "⚠️",
      "book"             => "📖",
      "white_check_mark" => "✅"
    },
    "bitbucket_server" => {
      "no_entry_sign"    => "\u274C", # ❌
      "warning"          => "\u26A0", # ⚠️
      "book"             => "\u2728", # ✨
      "white_check_mark" => "\u2705"  # ✅
    }
  }

  def initialize(template: "github")
    @template = (DATA.include? template) ? template : "github"
  end

  def map(emoji)
    emoji.delete! ":"
    DATA[template][emoji]
  end

  private

  attr_reader :template
end

EmojiMapper.new().map(":warning").to_s + " " + \
EmojiMapper.new(template: "bitbucket_server").map(":warning").to_s + " " + \
EmojiMapper.new(template: "bitbucket").map(":warning").to_s
```

Comment (JuanitoFatas) [inline lib/danger/helpers/comments_helper.rb]: > Would be nice to have a separate attribute like @emoji_mapper

I am not sure what you mean but your implementation looks solid!

**Comment by orta** [comment]
Cool - @JuanitoFatas wanna give this one more look over - seems good from me

File: lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb [hunk 5]
```diff
@@ -140,8 +140,10 @@ def head_commit
     end
 
     # @!group Bitbucket Server Misc
-    # Returns a list of HTML anchors for a file, or files in the head repository.
+    # Returns a list of Markdown links for a file, or files in the head repository.
```
Comment (JuanitoFatas) [inline lib/danger/danger_core/plugins/dangerfile_bitbucket_server_plugin.rb]: Thank you for also updating the docs.

**Comment by JuanitoFatas** [comment]
@krin-san Would you like to be a danger contributor (join the danger organization)? 😊

**Comment by krin-san** [comment]
@JuanitoFatas, sure, I'd like to. I could help with something in future =]

**Comment by orta** [comment]
Added! Welcome!

**Comment by JuanitoFatas** [comment]
Aw I invited you or @orta faster than me?

**Comment by krin-san** [comment]
@JuanitoFatas you was first :)

**Comment by krin-san** [comment]
It was a different invitations! 😂
@JuanitoFatas, you invited me to a repo contributors granting with a push action rights
@orta, you invited to the Danger github organization.
So I can't judge who was the first =)

**Comment by JuanitoFatas** [comment]
> @orta, you invited to the Danger github organization.
> So I can't judge who was the first =)

Ah I guess @orta is right. Please ignore mine and accept his, thank you and have a lovely day.

**Comment by orta** [comment]
:D

**Comment by orta** [comment]
more is more

**Pull request merged**